// @playwright/test@1.32.0/lib/runner/uiMode.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.runUIMode = runUIMode;
var _server = require("playwright-core/lib/server");
var _utils = require("playwright-core/lib/utils");
var _compilationCache = require("../common/compilationCache");
var _multiplexer = require("../reporters/multiplexer");
var _teleEmitter = require("../reporters/teleEmitter");
var _reporters = require("./reporters");
var _tasks = require("./tasks");
var _utilsBundle = require("../utilsBundle");
/**
 * Copyright Microsoft Corporation. All rights reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

class UIMode {
  constructor(config) {
    this._config = void 0;
    this._page = void 0;
    this._testRun = void 0;
    this.globalCleanup = void 0;
    this._testWatcher = void 0;
    this._originalStderr = void 0;
    this._config = config;
    process.env.PW_LIVE_TRACE_STACKS = '1';
    config._internal.configCLIOverrides.forbidOnly = false;
    config._internal.configCLIOverrides.globalTimeout = 0;
    config._internal.configCLIOverrides.repeatEach = 0;
    config._internal.configCLIOverrides.shard = undefined;
    config._internal.configCLIOverrides.updateSnapshots = undefined;
    config._internal.listOnly = false;
    config._internal.passWithNoTests = true;
    for (const project of config.projects) project._internal.deps = [];
    for (const p of config.projects) p.retries = 0;
    config._internal.configCLIOverrides.use = config._internal.configCLIOverrides.use || {};
    config._internal.configCLIOverrides.use.trace = {
      mode: 'on',
      sources: false
    };
    this._originalStderr = process.stderr.write.bind(process.stderr);
    this._installGlobalWatcher();
  }
  _installGlobalWatcher() {
    const projectDirs = new Set();
    for (const p of this._config.projects) projectDirs.add(p.testDir);
    let coalescingTimer;
    const watcher = _utilsBundle.chokidar.watch([...projectDirs], {
      ignoreInitial: true,
      persistent: true
    }).on('all', async event => {
      if (event !== 'add' && event !== 'change' && event !== 'unlink') return;
      if (coalescingTimer) clearTimeout(coalescingTimer);
      coalescingTimer = setTimeout(() => {
        this._dispatchEvent({
          method: 'listChanged'
        });
      }, 200);
    });
    return watcher;
  }
  async runGlobalSetup() {
    const reporter = await (0, _reporters.createReporter)(this._config, 'watch');
    const taskRunner = (0, _tasks.createTaskRunnerForWatchSetup)(this._config, reporter);
    reporter.onConfigure(this._config);
    const context = {
      config: this._config,
      reporter,
      phases: []
    };
    const {
      status,
      cleanup: globalCleanup
    } = await taskRunner.runDeferCleanup(context, 0);
    if (status !== 'passed') {
      await globalCleanup();
      return status;
    }
    this.globalCleanup = globalCleanup;
    return status;
  }
  async showUI() {
    this._page = await (0, _server.showTraceViewer)([], 'chromium', {
      app: 'watch.html',
      headless: (0, _utils.isUnderTest)() && process.env.PWTEST_HEADED_FOR_TEST !== '1'
    });
    if (!process.env.PWTEST_DEBUG) {
      process.stdout.write = chunk => {
        this._dispatchEvent({
          method: 'stdio',
          params: chunkToPayload('stdout', chunk)
        });
        return true;
      };
      process.stderr.write = chunk => {
        this._dispatchEvent({
          method: 'stdio',
          params: chunkToPayload('stderr', chunk)
        });
        return true;
      };
    }
    const exitPromise = new _utils.ManualPromise();
    this._page.on('close', () => exitPromise.resolve());
    let queue = Promise.resolve();
    this._page.exposeBinding('sendMessage', false, async (source, data) => {
      const {
        method,
        params
      } = data;
      if (method === 'exit') {
        exitPromise.resolve();
        return;
      }
      if (method === 'watch') {
        this._watchFiles(params.fileNames);
        return;
      }
      if (method === 'open' && params.location) {
        (0, _utilsBundle.open)('vscode://file/' + params.location).catch(e => this._originalStderr(String(e)));
        return;
      }
      if (method === 'resizeTerminal') {
        process.stdout.columns = params.cols;
        process.stdout.rows = params.rows;
        process.stderr.columns = params.cols;
        process.stderr.columns = params.rows;
        return;
      }
      if (method === 'stop') {
        this._stopTests();
        return;
      }
      queue = queue.then(() => this._queueListOrRun(method, params));
      await queue;
    });
    await exitPromise;
  }
  async _queueListOrRun(method, params) {
    if (method === 'list') await this._listTests();
    if (method === 'run') await this._runTests(params.testIds);
  }
  _dispatchEvent(message) {
    // eslint-disable-next-line no-console
    this._page.mainFrame().evaluateExpression(dispatchFuncSource, true, message).catch(e => this._originalStderr(String(e)));
  }
  async _listTests() {
    const listReporter = new _teleEmitter.TeleReporterEmitter(e => this._dispatchEvent(e));
    const reporter = new _multiplexer.Multiplexer([listReporter]);
    this._config._internal.listOnly = true;
    this._config._internal.testIdMatcher = undefined;
    const taskRunner = (0, _tasks.createTaskRunnerForList)(this._config, reporter, 'out-of-process');
    const context = {
      config: this._config,
      reporter,
      phases: []
    };
    (0, _compilationCache.clearCompilationCache)();
    reporter.onConfigure(this._config);
    const status = await taskRunner.run(context, 0);
    reporter.onExit({
      status
    });
  }
  async _runTests(testIds) {
    await this._stopTests();
    const testIdSet = testIds ? new Set(testIds) : null;
    this._config._internal.listOnly = false;
    this._config._internal.testIdMatcher = id => !testIdSet || testIdSet.has(id);
    const runReporter = new _teleEmitter.TeleReporterEmitter(e => this._dispatchEvent(e));
    const reporter = await (0, _reporters.createReporter)(this._config, 'ui', [runReporter]);
    const taskRunner = (0, _tasks.createTaskRunnerForWatch)(this._config, reporter);
    const context = {
      config: this._config,
      reporter,
      phases: []
    };
    (0, _compilationCache.clearCompilationCache)();
    reporter.onConfigure(this._config);
    const stop = new _utils.ManualPromise();
    const run = taskRunner.run(context, 0, stop).then(async status => {
      await reporter.onExit({
        status
      });
      this._testRun = undefined;
      this._config._internal.testIdMatcher = undefined;
      return status;
    });
    this._testRun = {
      run,
      stop
    };
    await run;
  }
  async _watchFiles(fileNames) {
    const files = new Set();
    for (const fileName of fileNames) {
      files.add(fileName);
      (0, _compilationCache.dependenciesForTestFile)(fileName).forEach(file => files.add(file));
    }
    const watchedFiles = [...files].sort();
    if (this._testWatcher && JSON.stringify(this._testWatcher.watchedFiles) === JSON.stringify(watchedFiles)) return;
    if (this._testWatcher) {
      if (this._testWatcher.collector.size) this._dispatchEvent({
        method: 'filesChanged',
        params: {
          fileNames: [...this._testWatcher.collector]
        }
      });
      clearTimeout(this._testWatcher.timer);
      this._testWatcher.watcher.close().then(() => {});
      this._testWatcher = undefined;
    }
    if (!watchedFiles.length) return;
    const collector = new Set();
    const watcher = _utilsBundle.chokidar.watch(watchedFiles, {
      ignoreInitial: true
    }).on('all', async (event, file) => {
      if (event !== 'add' && event !== 'change') return;
      (0, _compilationCache.collectAffectedTestFiles)(file, collector);
      if (this._testWatcher.timer) clearTimeout(this._testWatcher.timer);
      this._testWatcher.timer = setTimeout(() => {
        const fileNames = [...collector];
        collector.clear();
        this._dispatchEvent({
          method: 'filesChanged',
          params: {
            fileNames
          }
        });
      }, 250);
    });
    this._testWatcher = {
      watcher,
      watchedFiles,
      collector
    };
  }
  async _stopTests() {
    var _this$_testRun, _this$_testRun$stop, _this$_testRun2;
    (_this$_testRun = this._testRun) === null || _this$_testRun === void 0 ? void 0 : (_this$_testRun$stop = _this$_testRun.stop) === null || _this$_testRun$stop === void 0 ? void 0 : _this$_testRun$stop.resolve();
    await ((_this$_testRun2 = this._testRun) === null || _this$_testRun2 === void 0 ? void 0 : _this$_testRun2.run);
  }
}
const dispatchFuncSource = String(message => {
  window.dispatch(message);
});
async function runUIMode(config) {
  var _uiMode$globalCleanup;
  const uiMode = new UIMode(config);
  const status = await uiMode.runGlobalSetup();
  if (status !== 'passed') return status;
  await uiMode.showUI();
  return (await ((_uiMode$globalCleanup = uiMode.globalCleanup) === null || _uiMode$globalCleanup === void 0 ? void 0 : _uiMode$globalCleanup.call(uiMode))) || 'passed';
}
function chunkToPayload(type, chunk) {
  if (chunk instanceof Buffer) return {
    type,
    buffer: chunk.toString('base64')
  };
  return {
    type,
    text: chunk
  };
}
