// vitepress@1.0.0-alpha.61/dist/node/index.js
import { o as out, c as createMarkdownRenderer, g as grayMatter } from './serve-a6868a38.js';
export { S as ScaffoldThemeType, j as build, i as createServer, a as defineConfig, b as defineConfigWithTheme, d as defineLoader, k as init, r as resolveConfig, h as resolvePages, f as resolveSiteData, e as resolveUserConfig, l as scaffold, s as serve } from './serve-a6868a38.js';
import fs__default from 'fs';
import path from 'path';
import { normalizePath } from 'vite';
export { loadEnv } from 'vite';
import 'dns';
import 'node:process';
import 'assert';
import 'events';
import 'node:readline';
import 'stream';
import 'buffer';
import 'url';
import 'crypto';
import 'shiki';
import 'child_process';
import 'module';
import 'node:path';
import 'node:url';
import 'node:fs';
import 'tty';
import 'util';
import 'net';
import 'zlib';
import 'http';
import 'querystring';
import 'os';
import 'node:tty';
import 'constants';

function createContentLoader(pattern, {
  includeSrc,
  render,
  excerpt: renderExcerpt,
  transform
} = {}) {
  const config = global.VITEPRESS_CONFIG;
  if (!config) {
    throw new Error(
      "content loader invoked without an active vitepress process, or before vitepress config is resolved."
    );
  }
  if (typeof pattern === "string")
    pattern = [pattern];
  pattern = pattern.map((p) => normalizePath(path.join(config.root, p)));
  let md;
  const cache = /* @__PURE__ */ new Map();
  return {
    watch: pattern,
    async load(files) {
      if (!files) {
        files = (await out(pattern, {
          ignore: ["**/node_modules/**", "**/dist/**"]
        })).sort();
      }
      md = md || await createMarkdownRenderer(
        config.srcDir,
        config.markdown,
        config.site.base,
        config.logger
      );
      const raw = [];
      for (const file of files) {
        if (!file.endsWith(".md")) {
          continue;
        }
        const timestamp = fs__default.statSync(file).mtimeMs;
        const cached = cache.get(file);
        if (cached && timestamp === cached.timestamp) {
          raw.push(cached.data);
        } else {
          const src = fs__default.readFileSync(file, "utf-8");
          const { data: frontmatter, excerpt } = grayMatter(src, {
            excerpt: true
          });
          const url = "/" + normalizePath(path.relative(config.root, file)).replace(
            /\.md$/,
            config.cleanUrls ? "" : ".html"
          );
          const html = render ? md.render(src) : void 0;
          const renderedExcerpt = renderExcerpt ? excerpt && md.render(excerpt) : void 0;
          const data = {
            src: includeSrc ? src : void 0,
            html,
            frontmatter,
            excerpt: renderedExcerpt,
            url
          };
          cache.set(file, { data, timestamp });
          raw.push(data);
        }
      }
      return transform ? transform(raw) : raw;
    }
  };
}

export { createContentLoader, createMarkdownRenderer };
